File: objc/Tests/TestTransportRamanImage.m
/*
 * TestTransportRamanImage.m — Stage 5.3 of transport-spec v0.11
 * (Deferral 1).
 *
 * Exercises -[TTIOTransportWriter writeRamanImage:] + the matching
 * reader modality-dispatch path for modality=1 (Raman). The
 * IMAGE_HEADER carries a 16-byte modality_extras slot at its tail
 * containing two FLOAT64s: excitation_wavelength_nm + laser_power_mw.
 *
 * Cross-language parity:
 *   Java TransportRamanImageTest (commit f99ec47d)
 *   Python tests/test_transport_raman_image.py (commit 6abead73)
 *
 * SPDX-License-Identifier: Apache-2.0
 */
#import <Foundation/Foundation.h>
#import "Testing.h"

#import "Transport/TTIOTransportPacket.h"
#import "Transport/TTIOTransportReader.h"
#import "Transport/TTIOTransportReader+Internal.h"
#import "Transport/TTIOTransportWriter.h"
#import "Dataset/TTIOSpectralDataset.h"
#import "Image/TTIORamanImage.h"
#import "Image/TTIOIRImage.h"
#import "Image/TTIOMSImage.h"
#include <unistd.h>

static NSString *makeRamanTempPath(NSString *suffix)
{
    NSString *base = [NSString stringWithFormat:@"ttio_tr_raman_%d_%@.tio",
                       (int)getpid(), suffix];
    return [NSTemporaryDirectory() stringByAppendingPathComponent:base];
}

static uint16_t leU16Rm(const uint8_t *b)
{
    return (uint16_t)((uint32_t)b[0] | ((uint32_t)b[1] << 8));
}

static uint32_t leU32Rm(const uint8_t *b)
{
    return (uint32_t)b[0]
         | ((uint32_t)b[1] << 8)
         | ((uint32_t)b[2] << 16)
         | ((uint32_t)b[3] << 24);
}

static double leF64Rm(const uint8_t *b)
{
    uint64_t lo = (uint64_t)leU32Rm(b);
    uint64_t hi = (uint64_t)leU32Rm(b + 4);
    uint64_t bits = lo | (hi << 32);
    double d;
    memcpy(&d, &bits, 8);
    return d;
}

static TTIORamanImage *buildRamanFixture(void)
{
    const NSUInteger w = 3, h = 3, s = 4;
    NSMutableData *cube = [NSMutableData dataWithLength:w * h * s * sizeof(double)];
    double *p = (double *)cube.mutableBytes;
    for (NSUInteger y = 0; y < h; y++) {
        for (NSUInteger x = 0; x < w; x++) {
            NSUInteger pixelIdx = x + y * w;
            NSUInteger base = (y * w + x) * s;
            for (NSUInteger k = 0; k < s; k++) {
                p[base + k] = (double)(k + 1) * (double)pixelIdx;
            }
        }
    }
    NSMutableData *wn = [NSMutableData dataWithLength:s * sizeof(double)];
    double *wnp = (double *)wn.mutableBytes;
    for (NSUInteger i = 0; i < s; i++) wnp[i] = 500.0 + (double)i * 50.0;

    return [[TTIORamanImage alloc]
                initWithTitle:@"raman_fixture"
           isaInvestigationId:@""
              identifications:@[]
              quantifications:@[]
            provenanceRecords:@[]
                        width:w
                       height:h
               spectralPoints:s
                     tileSize:32
                   pixelSizeX:12.5
                   pixelSizeY:12.5
                  scanPattern:@"raster"
       excitationWavelengthNm:785.0
                 laserPowerMw:50.0
                         cube:cube
                  wavenumbers:wn];
}

// -------- 1. wire layout: modality=1 + 16-byte Raman extras tail ----------

static void testRamanImageWireLayout(void)
{
    TTIORamanImage *img = buildRamanFixture();

    NSMutableData *buf = [NSMutableData data];
    TTIOTransportWriter *w =
        [[TTIOTransportWriter alloc] initWithMutableData:buf];
    NSError *err = nil;
    PASS([w writeStreamHeaderWithFormatVersion:@"1.2"
                                          title:@"raman-wire"
                               isaInvestigation:@""
                                       features:@[]
                                      nDatasets:0
                                          error:&err],
         "5.3 raman wire: StreamHeader");
    PASS([w writeRamanImage:img error:&err],
         "5.3 raman wire: writeRamanImage emitted");
    PASS([w writeEndOfStreamWithError:&err],
         "5.3 raman wire: EndOfStream");

    TTIOTransportReader *r = [[TTIOTransportReader alloc] initWithData:buf];
    NSArray<TTIOTransportPacketRecord *> *records =
        [r readAllPacketsWithError:&err];
    // StreamHeader + IMAGE_HEADER + 9 IMAGE_PIXEL + EOI + EOS = 13
    PASS(records.count == 13,
         "5.3 raman wire: 13 packets total");

    NSData *hdr = records[1].payload;
    const uint8_t *hb = hdr.bytes;
    NSUInteger ho = 0;
    PASS(hb[ho] == 1u,
         "5.3 raman wire: modality == 1 (Raman)"); ho += 1;
    PASS(leU32Rm(&hb[ho]) == 3u, "5.3 raman wire: width == 3"); ho += 4;
    PASS(leU32Rm(&hb[ho]) == 3u, "5.3 raman wire: height == 3"); ho += 4;
    PASS(leU32Rm(&hb[ho]) == 4u, "5.3 raman wire: bins == 4"); ho += 4;
    PASS(leF64Rm(&hb[ho]) == 12.5, "5.3 raman wire: pixel_size_x"); ho += 8;
    PASS(leF64Rm(&hb[ho]) == 12.5, "5.3 raman wire: pixel_size_y"); ho += 8;
    PASS(hb[ho] == 0u, "5.3 raman wire: scan_pattern == 0 (raster)"); ho += 1;
    PASS(hb[ho] == 1u, "5.3 raman wire: axis_kind == 1 (wavenumber)"); ho += 1;
    uint32_t axisLen = leU32Rm(&hb[ho]); ho += 4;
    PASS(axisLen == 4u, "5.3 raman wire: axis_length == 4");
    for (NSUInteger i = 0; i < axisLen; i++) {
        double exp = 500.0 + (double)i * 50.0;
        PASS(leF64Rm(&hb[ho]) == exp,
             "5.3 raman wire: wavenumber byte-equal");
        ho += 8;
    }
    PASS(hb[ho] == 1u, "5.3 raman wire: is_continuous == 1"); ho += 1;
    uint16_t titleLen = leU16Rm(&hb[ho]); ho += 2; ho += titleLen;
    uint16_t isaLen = leU16Rm(&hb[ho]); ho += 2; ho += isaLen;
    uint16_t extrasLen = leU16Rm(&hb[ho]); ho += 2;
    PASS(extrasLen == 16u,
         "5.3 raman wire: modality_extras_length == 16 (8B excitation + 8B power)");
    double excitation = leF64Rm(&hb[ho]); ho += 8;
    double laserPower = leF64Rm(&hb[ho]); ho += 8;
    PASS(excitation == 785.0, "5.3 raman wire: excitation_wavelength_nm");
    PASS(laserPower == 50.0,  "5.3 raman wire: laser_power_mw");
    PASS(ho == hdr.length, "5.3 raman wire: no trailing bytes");
}

// -------- 2. round-trip via writeTtioToPath: -> reopen RamanImage ----------

static void testRamanImageRoundTrip(void)
{
    TTIORamanImage *img = buildRamanFixture();
    // Persist as a .tio so the writer can read it through the dataset.
    NSString *srcPath = makeRamanTempPath(@"rt-src");
    unlink([srcPath fileSystemRepresentation]);
    NSError *err = nil;
    PASS([img writeToFilePath:srcPath error:&err] && err == nil,
         "5.3 raman rt: source .tio written");

    TTIOSpectralDataset *ds = [TTIOSpectralDataset readFromFilePath:srcPath
                                                                error:&err];
    PASS(ds != nil && ds.ramanImage != nil,
         "5.3 raman rt: source dataset carries RamanImage");

    NSMutableData *buf = [NSMutableData data];
    TTIOTransportWriter *w =
        [[TTIOTransportWriter alloc] initWithMutableData:buf];
    PASS([w writeStreamHeaderWithFormatVersion:@"1.2"
                                          title:@"raman_fixture"
                               isaInvestigation:@""
                                       features:@[TTIOTransportV011Feature]
                                      nDatasets:0
                                          error:&err],
         "5.3 raman rt: StreamHeader");
    PASS([w writeRamanImage:ds.ramanImage error:&err],
         "5.3 raman rt: writeRamanImage");
    PASS([w writeEndOfStreamWithError:&err], "5.3 raman rt: EndOfStream");

    NSString *rtPath = makeRamanTempPath(@"rt");
    unlink([rtPath fileSystemRepresentation]);
    TTIOTransportReader *r = [[TTIOTransportReader alloc] initWithData:buf];
    PASS([r writeTtioToPath:rtPath error:&err] && err == nil,
         "5.3 raman rt: writeTtioToPath materialised");

    TTIOSpectralDataset *rt = [TTIOSpectralDataset readFromFilePath:rtPath
                                                                error:&err];
    PASS(rt != nil && rt.ramanImage != nil,
         "5.3 raman rt: round-tripped dataset carries RamanImage");
    TTIORamanImage *back = rt.ramanImage;
    PASS(back.width == img.width
        && back.height == img.height
        && back.spectralPoints == img.spectralPoints,
         "5.3 raman rt: dims match");
    PASS(back.excitationWavelengthNm == img.excitationWavelengthNm,
         "5.3 raman rt: excitation_wavelength_nm round-trip");
    PASS(back.laserPowerMw == img.laserPowerMw,
         "5.3 raman rt: laser_power_mw round-trip");
    PASS([back.cube isEqualToData:img.cube],
         "5.3 raman rt: cube byte-equal");
    PASS([back.wavenumbers isEqualToData:img.wavenumbers],
         "5.3 raman rt: wavenumbers byte-equal");
    // The materialised .tio carries only Raman; MS/IR remain absent.
    PASS(rt.irImage == nil,
         "5.3 raman rt: irImage stays nil");

    [ds closeFile];
    [rt closeFile];
    unlink([srcPath fileSystemRepresentation]);
    unlink([rtPath fileSystemRepresentation]);
}

// -------- 3. unknown modality (99) is logged + skipped ---------------------

static void testReaderSkipsUnknownModality(void)
{
    // Hand-roll the IMAGE_HEADER for modality=99 + a single dummy
    // pixel + END_OF_IMAGE via the writer's low-level emitPacket
    // API. The writer doesn't expose emitPacket directly — synthesise
    // by patching a Raman stream's IMAGE_HEADER modality byte to 99.
    TTIORamanImage *img = buildRamanFixture();

    NSMutableData *buf = [NSMutableData data];
    TTIOTransportWriter *w =
        [[TTIOTransportWriter alloc] initWithMutableData:buf];
    NSError *err = nil;
    PASS([w writeStreamHeaderWithFormatVersion:@"1.2"
                                          title:@"unknown_mod"
                               isaInvestigation:@""
                                       features:@[]
                                      nDatasets:0
                                          error:&err],
         "5.3 unknown: StreamHeader");
    PASS([w writeRamanImage:img error:&err],
         "5.3 unknown: wrote Raman image (will patch modality)");
    PASS([w writeEndOfStreamWithError:&err], "5.3 unknown: EndOfStream");

    // Locate the IMAGE_HEADER payload's first byte (the modality)
    // and flip it from 1 to 99. The IMAGE_HEADER payload starts
    // with `modality(u8)=1 + width(u32 LE)=3 + height(u32 LE)=3 +
    // spectrum_bins(u32 LE)=4`, so the byte pattern is
    // 01 03 00 00 00 03 00 00 00 04 00 00 00 ...
    NSMutableData *mut = [buf mutableCopy];
    uint8_t *p = mut.mutableBytes;
    NSUInteger n = mut.length;
    NSUInteger headerOff = NSNotFound;
    for (NSUInteger i = 0; i + 13 <= n; i++) {
        if (p[i]   == 0x01 && p[i+1] == 0x03 && p[i+2] == 0x00
         && p[i+3] == 0x00 && p[i+4] == 0x00 && p[i+5] == 0x03
         && p[i+6] == 0x00 && p[i+7] == 0x00 && p[i+8] == 0x00
         && p[i+9] == 0x04 && p[i+10] == 0x00 && p[i+11] == 0x00
         && p[i+12] == 0x00) {
            headerOff = i;
            break;
        }
    }
    PASS(headerOff != NSNotFound,
         "5.3 unknown: located Raman IMAGE_HEADER modality byte");
    if (headerOff != NSNotFound) {
        p[headerOff] = 99u;
    }

    NSString *outPath = makeRamanTempPath(@"unknown");
    unlink([outPath fileSystemRepresentation]);
    TTIOTransportReader *r = [[TTIOTransportReader alloc] initWithData:mut];
    BOOL ok = [r writeTtioToPath:outPath error:&err];
    PASS(ok && err == nil,
         "5.3 unknown: reader accepts stream (skipping unknown modality)");

    // The materialised .tio has no Raman image (the unknown-modality
    // block was skipped).
    TTIOSpectralDataset *rt = [TTIOSpectralDataset readFromFilePath:outPath
                                                                error:&err];
    PASS(rt != nil, "5.3 unknown: re-opened materialised");
    PASS(rt.ramanImage == nil,
         "5.3 unknown: unknown-modality stream produces no RamanImage");
    PASS(rt.irImage == nil,
         "5.3 unknown: unknown-modality stream produces no IRImage");
    // msImage's placeholder semantics — check width==0.
    PASS(rt.msImage == nil || rt.msImage.width == 0,
         "5.3 unknown: unknown-modality stream produces no MSImage cube");

    [rt closeFile];
    unlink([outPath fileSystemRepresentation]);
}

// -------- entry point ------------------------------------------------------

void testTransportRamanImage(void);
void testTransportRamanImage(void)
{
    testRamanImageWireLayout();
    testRamanImageRoundTrip();
    testReaderSkipsUnknownModality();
}

File: python/tests/test_transport_raman_image.py
"""v0.11 Task 5.3 (Deferral 1, transport-spec §4.16): exercise the
IMAGE pipeline for :class:`~ttio.RamanImage` (modality=1) on
:meth:`TransportWriter.write_raman_image` and the matching reader
modality-dispatch path.

The IMAGE_HEADER carries a ``modality_extras`` slot at its tail with
the Raman-specific fields ``excitation_wavelength_nm +
laser_power_mw`` (two FLOAT64, 16 bytes total).

The shared header otherwise matches MS (modality=0) verbatim; each
pixel rides as a continuous-mode IMAGE_PIXEL whose payload is a
dense vector of ``spectrum_bins`` float64 intensities. The shared
axis on the IMAGE_HEADER is the Raman wavenumbers vector
(``axis_kind=1``).

Python parity for Java's :class:`TransportRamanImageTest` (commit
``f99ec47d``).

SPDX-License-Identifier: Apache-2.0
"""
from __future__ import annotations

import io
import struct
from pathlib import Path

import numpy as np

from ttio import SpectralDataset
from ttio.enums import ImageKind
from ttio.raman_image import RamanImage
from ttio.transport.codec import TransportReader, TransportWriter
from ttio.transport.packets import PacketType


def _build_raman_fixture(target: Path) -> Path:
    """Build a 3x3x4 RamanImage .tio file (mirrors Java)."""
    w, h, s = 3, 3, 4
    cube = np.zeros((h, w, s), dtype=np.float64)
    for y in range(h):
        for x in range(w):
            for k in range(s):
                cube[y, x, k] = (k + 1.0) * (x + y * w)
    wn = np.array([500.0 + i * 50.0 for i in range(s)], dtype=np.float64)
    img = RamanImage(
        width=w, height=h, spectral_points=s,
        intensity=cube, wavenumbers=wn,
        pixel_size_x=12.5, pixel_size_y=12.5, scan_pattern="raster",
        excitation_wavelength_nm=785.0, laser_power_mw=50.0,
        title="raman_fixture", isa_investigation_id="",
    )
    SpectralDataset.write_minimal(
        target,
        title="raman_fixture",
        isa_investigation_id="",
        runs={},
        raman_image=img,
    )
    return target


def test_write_raman_image_emits_header_pixels_eoi_with_modality_1(
    tmp_path: Path,
) -> None:
    """Writer emits IMAGE_HEADER + N IMAGE_PIXEL + END_OF_IMAGE with
    modality=1 (Raman) and the modality_extras slot at the tail of
    the IMAGE_HEADER carrying the two Raman FLOAT64 fields."""
    src = _build_raman_fixture(tmp_path / "raman.tio")
    tis = tmp_path / "raman.tis"

    with SpectralDataset.open(src) as ds:
        assert ds.image_for_kind(ImageKind.RAMAN) is not None, (
            "fixture precondition: dataset must carry a RamanImage"
        )
        out = io.BytesIO()
        with TransportWriter(out) as w:
            w.write_stream_header(
                format_version="1.2",
                title=ds.title or "",
                isa_investigation=ds.isa_investigation_id or "",
                features=[],
                n_datasets=0,
            )
            w.write_raman_image(ds.image_for_kind(ImageKind.RAMAN))
            w.write_end_of_stream()
        tis.write_bytes(out.getvalue())

    r = TransportReader(io.BytesIO(tis.read_bytes()))
    records = r.records_for_test()
    # 1 StreamHeader + 1 IMAGE_HEADER + 9 IMAGE_PIXEL + 1 EOI + 1 EOS
    assert len(records) == 13, (
        f"expected StreamHeader + IMAGE_HEADER + 9 pixels + EOI + EOS, "
        f"got {len(records)}"
    )
    assert records[0].header.packet_type == int(PacketType.STREAM_HEADER)
    assert records[1].header.packet_type == int(PacketType.IMAGE_HEADER)
    for i in range(9):
        assert records[2 + i].header.packet_type == int(PacketType.IMAGE_PIXEL)
    assert records[11].header.packet_type == int(PacketType.END_OF_IMAGE)
    assert records[12].header.packet_type == int(PacketType.END_OF_STREAM)

    hdr = records[1].payload
    off = 0
    (modality,) = struct.unpack_from("<B", hdr, off); off += 1
    (width,) = struct.unpack_from("<I", hdr, off); off += 4
    (height,) = struct.unpack_from("<I", hdr, off); off += 4
    (bins,) = struct.unpack_from("<I", hdr, off); off += 4
    (px_x,) = struct.unpack_from("<d", hdr, off); off += 8
    (px_y,) = struct.unpack_from("<d", hdr, off); off += 8
    (scan_pat,) = struct.unpack_from("<B", hdr, off); off += 1
    (axis_kind,) = struct.unpack_from("<B", hdr, off); off += 1
    (axis_len,) = struct.unpack_from("<I", hdr, off); off += 4
    axis = [struct.unpack_from("<d", hdr, off + 8 * i)[0]
            for i in range(axis_len)]
    off += 8 * axis_len
    (continuous,) = struct.unpack_from("<B", hdr, off); off += 1
    (title_len,) = struct.unpack_from("<H", hdr, off); off += 2
    off += title_len
    (isa_len,) = struct.unpack_from("<H", hdr, off); off += 2
    off += isa_len
    (extras_len,) = struct.unpack_from("<H", hdr, off); off += 2

    assert modality == 1, "RamanImage maps to modality 1"
    assert width == 3
    assert height == 3
    assert bins == 4
    assert px_x == 12.5
    assert px_y == 12.5
    assert scan_pat == 0, "raster maps to scan_pattern 0"
    assert axis_kind == 1, "RamanImage axis_kind is wavenumber=1"
    assert axis_len == 4
    for i, v in enumerate(axis):
        assert v == 500.0 + i * 50.0
    assert continuous == 1, "fixture is continuous mode"
    assert extras_len == 16, (
        "modality=1 extras = 8B excitation + 8B laser_power = 16B"
    )
    (exc,) = struct.unpack_from("<d", hdr, off); off += 8
    (laser,) = struct.unpack_from("<d", hdr, off); off += 8
    assert exc == 785.0
    assert laser == 50.0

    (pixel_count,) = struct.unpack_from("<I", records[11].payload, 0)
    assert pixel_count == 9, "pixel_count_seen = width*height"


def test_raman_image_round_trips_via_write_dataset_materialize(
    tmp_path: Path,
) -> None:
    """End-to-end: write_dataset → read_to_dataset → re-read; the
    round-tripped dataset carries the RamanImage with all fields
    byte-equal to the original."""
    src = _build_raman_fixture(tmp_path / "src.tio")
    tis = tmp_path / "src.tis"
    rt = tmp_path / "rt.tio"

    with SpectralDataset.open(src) as ds:
        with tis.open("wb") as out, TransportWriter(out) as w:
            w.write_dataset(ds)

    with TransportReader(io.BytesIO(tis.read_bytes())) as r:
        materialised = r.read_to_dataset(output_path=rt)
        materialised.close()

    with SpectralDataset.open(src) as a, SpectralDataset.open(rt) as b:
        img_a = a.image_for_kind(ImageKind.RAMAN)
        img_b = b.image_for_kind(ImageKind.RAMAN)
        assert img_a is not None, "source must carry a RamanImage"
        assert img_b is not None, "round-tripped dataset must carry a RamanImage"
        assert img_a.width == img_b.width
        assert img_a.height == img_b.height
        assert img_a.spectral_points == img_b.spectral_points
        assert img_a.pixel_size_x == img_b.pixel_size_x
        assert img_a.pixel_size_y == img_b.pixel_size_y
        assert img_a.scan_pattern == img_b.scan_pattern
        assert img_a.excitation_wavelength_nm == img_b.excitation_wavelength_nm
        assert img_a.laser_power_mw == img_b.laser_power_mw
        np.testing.assert_array_equal(img_a.wavenumbers, img_b.wavenumbers)
        np.testing.assert_array_equal(img_a.intensity, img_b.intensity)


def test_reader_skips_unknown_modality(tmp_path: Path) -> None:
    """Reader logs + skips an unknown modality byte (modality=99)
    rather than aborting the stream. The trailing END_OF_STREAM is
    still observed; the rest of the stream remains parseable."""
    # Synthesize a stream with an IMAGE_HEADER (modality=99, 1x1x1)
    # + 1 IMAGE_PIXEL + END_OF_IMAGE + END_OF_STREAM.
    tis = tmp_path / "unknown_mod.tis"
    sink = io.BytesIO()

    # IMAGE_HEADER payload (modality=99, 1+4+4+4+8+8+1+1+4+(0)+1+2+0+2+0+2+0
    # = 42 bytes — fixed prefix only, no axis values, no title/isa/
    # extras bodies).
    hbuf = b"".join((
        struct.pack("<B", 99),       # modality (unknown)
        struct.pack("<I", 1),        # width
        struct.pack("<I", 1),        # height
        struct.pack("<I", 1),        # spectrum_bins
        struct.pack("<d", 1.0),      # pixel_size_x
        struct.pack("<d", 1.0),      # pixel_size_y
        struct.pack("<B", 0),        # scan_pattern
        struct.pack("<B", 0),        # axis_kind
        struct.pack("<I", 0),        # axis_length
        # axis omitted (length 0)
        struct.pack("<B", 1),        # is_continuous
        struct.pack("<H", 0),        # title_length
        struct.pack("<H", 0),        # isa_id_length
        struct.pack("<H", 0),        # modality_extras_length
    ))
    assert len(hbuf) == 42

    # IMAGE_PIXEL (continuous, 1 float64 intensity).
    pbuf = b"".join((
        struct.pack("<I", 0),        # x
        struct.pack("<I", 0),        # y
        struct.pack("<B", 1),        # precision (FLOAT64)
        struct.pack("<B", 0),        # compression (NONE)
        struct.pack("<I", 8),        # payload_length
        struct.pack("<d", 42.0),     # single intensity
    ))

    # END_OF_IMAGE payload.
    ebuf = struct.pack("<I", 1)

    with TransportWriter(sink) as w:
        w.write_stream_header(
            format_version="1.2",
            title="unknown_mod",
            isa_investigation="",
            features=[],
            n_datasets=0,
        )
        # Synthesise unusual packets via the private _emit hook;
        # Python writer has no public emit_raw_packet sibling.
        w._emit(PacketType.IMAGE_HEADER, hbuf)
        w._emit(PacketType.IMAGE_PIXEL, pbuf)
        w._emit(PacketType.END_OF_IMAGE, ebuf)
        w.write_end_of_stream()
    tis.write_bytes(sink.getvalue())

    # materializeTo must NOT throw on the unknown modality; the
    # image block is silently dropped.
    rt = tmp_path / "rt.tio"
    with TransportReader(io.BytesIO(tis.read_bytes())) as r:
        materialised = r.read_to_dataset(output_path=rt)
        materialised.close()
    with SpectralDataset.open(rt) as ds:
        assert ds.image_for_kind(ImageKind.MS) is None, (
            "unknown-modality stream must produce no MSImage"
        )
        assert ds.image_for_kind(ImageKind.RAMAN) is None, (
            "unknown-modality stream must produce no RamanImage"
        )
        assert ds.image_for_kind(ImageKind.IR) is None, (
            "unknown-modality stream must produce no IRImage"
        )
